### PR TITLE
Refactors Azure Blob Storage destination

### PR DESCRIPTION
**Pull request by Jsalz2000**
# Summary
This pull request refactors/adds an Azure Blob Storage implementation for twindb-backup. Removes a majority of the existing code for Azure Blob, as it was overly complex, and wasn't already integrated with the twindb-backup command.

This feature will allow users to backup and restore directly from an Azure Blob Storage Account.

# Config
```ini
[az]
connection_string = "DefaultEndpointsProtocol=https;AccountName=ACCOUNT_NAME;AccountKey=ACCOUNT_KEY;EndpointSuffix=core.windows.net"
container_name = twindb-backups
remote_path = /backups/mysql # optional
```

# Examples

## twindb-backup ls
```bash
bash-5.1# twindb-backup ls
2024-09-03 14:41:08,152: INFO: ls.list_available_backups():25: Destination <twindb_backup.destination.az.AZ object at 0xffff8533ec70>
2024-09-03 14:41:09,655: INFO: ls._print_media_type():44: mysql hourly copies:
coolhostname/hourly/mysql/mysql-2024-09-03_14_37_20.xbstream.gz
2024-09-03 14:41:10,922: INFO: ls._print_binlog():52: Binary logs:
coolhostname/binlog/binlog.000001.gz
coolhostname/binlog/binlog.000002.gz
coolhostname/binlog/binlog.000003.gz
coolhostname/binlog/binlog.000004.gz
```

## twindb-backup verify
```bash
bash-5.1# twindb-backup verify mysql coolhostname/hourly/mysql/mysql-2024-09-03_14_37_20.xbstream.gz
2024-09-03 14:43:29,295: INFO: restore.restore_from_mysql():320: Restoring MySQLCopy(coolhostname/hourly/mysql/mysql-2024-09-03_14_37_20.xbstream.gz) = {
    "backup_finished": 1725374247,
    "backup_started": 1725374240,

.....

2024-09-03 14:43:36,148: INFO: restore.restore_from_mysql():429: Then you can start MySQL normally.
{
    "backup_copy": "coolhostname/hourly/mysql/mysql-2024-09-03_14_37_20.xbstream.gz",
    "restore_time": 6.85528564453125,
    "success": true
}
```

## Azure Storage Account
Screenshot 2024-09-03 at 9 40 11 AM

### CODE REVIEW

File: docs/usage.rst [hunk 1]
```diff
@@ -47,7 +47,7 @@ Backup Destination
 The ``[destination]`` section specifies where to store backup copies.
 
 ``backup_destination`` can be either ``ssh`` (if you want to store backups on a remote SSH server),
-or ``s3`` (if you want to store backups in Amazon S3), or ``gsc`` (if the backup should be stored in Google Cloud).
+``s3`` (if you want to store backups in Amazon S3), ``az`` (if the backup should be stored in Azure Blob Storage), or ``gsc`` (if the backup should be stored in Google Cloud).
```
Comment (ernstae) [inline docs/usage.rst:50]: I think there's an original typo here -- in that it should read gcs and not gsc

**Review by ernstae — COMMENTED** [review]
This looks very complete and I'm excited to kick the tires and test it out. In the short-term, I caught a typo that was already in the documentation codebase that could be fixed.  


This, overall looks like complete coverage of the Azure Blob functionality and well-structured Python, James.

File: twindb_backup/destination/az.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                client.create_container()
+        except builtins.Exception as err:
+            LOG.error(f"Failed to validate or create container. Error: {type(err).__name__}, Reason: {err}")
+            raise err
+
+        return client
+
+    def render_path(self, path: str) -> str:
+        """Renders the absolute path for the Azure Blob Storage Destination
+
```
Comment (Jsalz2000) [inline twindb_backup/destination/az.py:89]: Missing arg in docstring

File: twindb_backup/destination/az.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        LOG.debug(f"Attempting to delete blob: {self.render_path(path)}")
+        try:
+            self._container_client.delete_blob(self.render_path(path))
+        except builtins.Exception as err:
+            LOG.error(f"Failed to delete blob {self.render_path(path)}. Error: {type(err).__name__}, Reason: {err}")
+            raise err
+
+    @contextmanager
+    def get_stream(self, copy):
```
Comment (Jsalz2000) [inline twindb_backup/destination/az.py:131]: Missing BaseCopy type hint and return type

File: docs/usage.rst (hunk 1 above)
Comment (Jsalz2000) [reply docs/usage.rst]: resolved

File: twindb_backup/destination/az.py (hunk 2 above)
Comment (Jsalz2000) [reply twindb_backup/destination/az.py:90]: resolved

File: twindb_backup/destination/az.py (hunk 3 above)
Comment (Jsalz2000) [reply twindb_backup/destination/az.py]: resolved

**Pull request merged**